![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesor: Pablo Badilla, Diego Cortez
- Auxiliar: Melanie Peña Torres, Valentina Rojas Osorio
- Ayudante: Javiera Arévalo, Tamara Carrasco, Ignacio Reyes

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Tomás Vallejo Cerda
- Estudiante n°2:

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

## Reglas

El proyecto debe realizarse al final del curso y su objetivo es poner a prueba el conocimiento de todas las unidades del curso. Este consta de 3 partes: **Informe de Análisis** (30%), **Construcción del Modelo** (40%) y **Despliegue del Modelo** (30%). Este notebook corresponde a la **Parte 1 - Informe de Análisis**.

### **Requisitos del proyecto**
- **Grupos**: Formar equipos de **2 personas**. 
- **Plazo de entrega**: Hasta el Viernes 10 de Julio
- **Consultas**: Cualquier duda fuera del horario de clases debe ser planteada en el foro correspondiente. Los mensajes enviados al equipo docente serán respondidos únicamente por este medio. Por favor, revisen las respuestas anteriores en el foro antes de realizar nuevas consultas.
- **Plagio**: La copia o reutilización no autorizada de trabajos de otros grupos está **estrictamente prohibida**. El incumplimiento de esta norma implicará la anulación inmediata del proyecto y una posible sanción académica.
- **Material permitido**: Pueden usar cualquier material del curso, ya sea notas, lecturas, códigos, o referencias proporcionadas por los docentes, que consideren útil para el desarrollo del proyecto.

---

## 📊 Distribución de Puntaje

Esta parte se evalúa en **escala de 1.0 a 7.0**, donde la nota se calcula como:

$$\text{Nota} = 1.0 + \text{puntaje obtenido}$$

El puntaje total repartible es de **6.0 puntos**, distribuidos así:

| Sección | Detalle | Puntaje |
| ------- | ------- | ------: |
| **Sección 1 - Entendimiento del Problema** | Preguntas conceptuales (P1–P7) | 2.0 |
| **Sección 2 · Subsección 1 - Datos Tabulares** | Código (1.0) + Preguntas (1.6) | 2.6 |
| **Sección 2 · Subsección 2 - Embeddings** | Código (0.6) + Preguntas (0.8) | 1.4 |
| **Total** | | **6.0** |

Cada pregunta y tarea indica su puntaje entre paréntesis, p. ej. _(0.3 pts)_.


Esta nota corresponde a la nota de la **Parte 1** que equivale a un **30%** de la nota del proyecto final. Para obtener la nota final, las notas individuales de cada parte se ponderan de la siguiente manera.

| Parte              | Ponderación |
|--------------------|------------:|
| **1 — Análisis**   | **30%**     |
| 2 — Modelo         | 40%         |
| 3 — Despliegue     | 30%         |

---

## 📖 Introducción 

<p align="center">
    <img src="./chaucherapp.jpg" alt="Alt Text" width="700">
</p>

La app chilena **ChaucherApp** ha emergido recientemente como una de las nuevas apps *FinTech* que apuesta por dominar en un competitivo mercado. Esta app ofrece una billetera digital que permite hacer transferencias, tarjetas digitales que permiten comprar, y servicios financieros como préstamos, seguros e inversiones.

Sin duda es sorprendente que **ChaucherApp** sea uno de los líderes de la industria fintech, siendo una empresa tan joven. En las palabras del CEO de la empresa, el Sr. Querty _'El gran éxito de ChaucherApp se debe a que entendimos la importancia de escuchar y responder eficientemente al cliente'_. En efecto, **ChaucherApp** fue diseñada desde el comienzo con una robusta arquitectura digital relacionada al **Soporte al cliente**.

Durante años, **ChaucherApp** ha logrado dar respuesta satisfactoria a cientos de tickets de soporte al cliente al mes, lo que le ha permitido inspirar confianza y prestigio. Por supuesto, esto también ha generado una gran cantidad de datos que pueden ser útiles para las siguientes etapas de crecimiento de la empresa. Y es que, a medida que el mercado se hace más competitivo, se hace más necesario aumentar la eficiencia de los procesos de respuesta de tickets, por lo que el Sr. Querty ha optado por **automatizar el proceso** incorporando inteligencia artificial.

Luego de intensa planificación ejecutiva y jornadas de analítica, han decidido que el primer paso para automatizar será el de desarrollar un modelo de **priorización de tickets** de soporte al cliente utilizando **modelos de lenguaje** y **aprendizaje de máquinas**. De esta manera podrán optimizar recursos humanos mientras se desarrollan otras herramientas.

Para lograr esta tarea, te han encargado analizar datos históricos de tickets de soporte al cliente que ya han sido manualmente priorizados, presentes en el archivo `data/tickets.parquet`. En él, encontrarás tickets con el contenido del ticket, categoría, canal, fecha y priorización, que puede ser `Baja`, `Media`, `Alta` o `Crítica`. Tu objetivo es utilizar esa data y los medios, algoritmos y estrategias que estimes convenientes para construir un **modelo clasificador** que sea capaz de predecir esta categoría en base al texto y atributos del ticket.

Con esta información, tus tareas a realizar son primero **analizar y entender los datos**, luego **construir un modelo de inferencia**, y finalmente **implementar ese modelo** para que el equipo de ejecutivos de ChaucherApp pueda utilizarlo.

---

# Sección 1 - Entendimiento del Problema

_(Total: 2.0 pts)_

El objetivo de esta sección es entender cuál será el problema que tendrás que resolver, incluyendo tanto la definición misma del problema (i.e., si es clasificación binaria, clasificación multiclase, regresión, clustering, otros...) como también las métricas, tanto de negocio como de evaluación, que usarás para medir el éxito de tu solución.

Para esto, primero responde las siguientes preguntas:


1. ¿Qué tipo de problema vamos a resolver? ¿Clasificación? ¿Regresión? ¿Clustering? ¿Reducción de dimensionalidad? ¿Ranking? ¿Recomendación? ¿Por qué? (utiliza la definición de cada uno de estos tipos de problemas más el enunciado para fundamentar). _(0.3 pts)_
2. ¿Por qué el negocio te solicitaría resolver este problema? ¿Por qué es importante? _(0.2 pts)_
3. ¿Por qué el negocio solicita generar un modelo predictivo para resolver este problema? ¿Por qué no lo pueden hacer, por ejemplo, por medio de código más palabras claves? _(0.2 pts)_
4. ¿Cuáles serían las métricas de evaluación relevantes para el negocio que buscarás optimizar? Aquí tienes la libertad de fundamentar lo que se te ocurra en base a tu entendimiento del negocio. Nombra al menos 3 y justifica la elección. _(0.3 pts)_
5. ¿Cuáles serían las métricas de evaluación de ML importantes para este problema? Para responder esto, primero da la interpretación de qué significa cada métrica (R^2, RMSE, MSE, etc.) / (Accuracy, Precision, Recall, F1) respecto a un problema binario Correcto / Incorrecto. Importante: Debes responder con la interpretación de qué significa cada una de estas métricas respecto al problema binario. _(0.5 pts)_
6. Luego, a partir de las interpretaciones anteriores de métricas más el conocimiento del negocio, elige cuál métrica es la más apropiada para cada clase en base a las prioridades del negocio. _(0.3 pts)_
7. Por último, ¿qué es la agregación macro? ¿Y la weighted? Y, ¿qué tipo de agregación utilizarías para promediar estas métricas de evaluación? _(0.2 pts)_

**Responde aquí**

1. Es un problema de clasificación multiclase supervisada. Es clasificación porque hay que asignar cada ticket a una categoría, no predecir un número (regresión) ni descubrir grupos sin etiquetas (clustering). Es multiclase porque `Nivel_Prioridad` tiene cuatro valores (`Baja`, `Media`, `Alta`, `Critica`), no dos. Y es supervisado porque el enunciado dice que los tickets ya vienen priorizados manualmente, es decir, tenemos la etiqueta real para entrenar el modelo. No es ranking ni recomendación, ya que aunque las prioridades tienen un orden, lo que se pide es etiquetar cada ticket por separado, no ordenarlos entre sí ni recomendarle algo a un usuario.

2. Porque priorizar tickets a mano no escala, como nos comentan a medida que crecen los usuarios de ChaucherApp, crecen los tickets, y el equipo de soporte tiene recursos limitados. Es importante porque un ticket crítico mal priorizado (por ejemplo, un fraude o un problema con dinero) puede tardar en atenderse igual que uno de baja importancia, generando pérdidas económicas y de confianza, donde en este tipo de negocios la confianza del cliente es lo central.

3. Porque el lenguaje de los tickets puede ser muy variado, donde la misma urgencia se puede expresar con palabras completamente distintas, con errores de tipeo, sarcasmo por enojo, o combinando el texto con otras señales (canal, tipo de cuenta, antigüedad). Las reglas de palabras clave son rígidas y se rompen fácil apenas alguien escribe distinto a lo esperado. Por otro lado, un modelo aprende patrones directamente de los datos históricos y generaliza mejor a casos nuevos, además de poder combinar texto con variables numéricas y categóricas a la vez, algo que un sistema de palabras clave no hace de forma natural.

4. Tres métricas de negocio relevantes:
   - **Cumplimiento de SLA en tickets críticos**: qué porcentaje de tickets críticos se atienden dentro del tiempo comprometido, porque ahí es donde más duele fallar.
   - **Ahorro en horas-hombre**: cuánto tiempo del equipo se libera al no tener que priorizar manualmente cada ticket.
   - **Retención de clientes**: un cliente que vive una mala experiencia con un ticket urgente mal priorizado tiene más probabilidad de irse a la competencia.

5. Como nuestro problema es de clasificación, las métricas de regresión no aplican, las realmente relevantes para este problema son las siguientes:

   - **Accuracy**: porcentaje de predicciones totales que acertó. Es la más intuitiva, pero poco confiable cuando hay desbalance de clases, que podría ser exactamente nuestro caso.
   - **Precision**: de todo lo que el modelo predijo como "Correcto", cuánto realmente lo era. Mide qué tan confiable es cuando el modelo dice que sí.
   - **Recall**: de todo lo que realmente era "Correcto", cuánto logró detectar el modelo. Mide qué tan bien no se le escapan los casos que sí importaban.
   - **F1**: combina precision y recall en un solo número mediante su media armónica, útil cuando quieres balancear ambas sin priorizar una a ciegas.


6. Para `Critica` lo más importante es el recall, ya que no podemos permitirnos que un ticket realmente crítico pase desapercibido como si fuera de baja prioridad, el costo de ese error es alto. Para `Baja` importa más la precision, no queremos gastar recursos tratando tickets sin urgencia como si fueran urgentes. Para `Media` y `Alta`, donde el costo de equivocarse en cualquier dirección es más parejo, tiene sentido usar F1 como balance entre ambas.

7. La agregación macro calcula la métrica por separado en cada clase y después saca el promedio simple, dándole el mismo peso a todas las clases sin importar cuántos ejemplos tenga cada una. La agregación weighted también promedia por clase, pero pondera según la cantidad de ejemplos de cada una, así que las clases mayoritarias pesan más en el resultado final. Acá a priori conviene usar macro, porque `Critica` es la clase más chica y la que más nos importa no descuidar, ya que si usáramos weighted, un mal desempeño en `Critica` quedaría opacado por el buen desempeño en `Media` y `Baja`, que son mayoría.

---

# Sección 2 - Análisis de Datos

## Subsección 1 - Datos Tabulares

_(Total: 2.6 pts — código 1.0 + preguntas 1.6)_


La idea de esta segunda sección es que leas las fuentes de datos y las explores.


Partiremos primero con `tickets.parquet`. Utiliza la función de I/O asociada a este tipo de archivo para cargar los datos.

In [ ]:
import pandas as pd

df_tickets = pd.read_parquet("data/tickets.parquet")
df_tickets.head()

,Id_Ticket,Asunto_Ticket,Contenido_Ticket,N_Caracteres_Ticket,Canal_Ticket,Categoría_Problema,Usuario-Tipo_de_Cuenta,Usuario-Antiguedad_Cuenta_Dias,Fecha_Envío,Tiempo_de_Resolución_Horas,Nivel_Prioridad
0,TCK-19267,Error al iniciar sesión en nuevo celular por d...,hola oye lo que pasa es que me cambie de celu ...,531,Whatsapp,Cuenta,Premium,159,2024-01-02,14,Baja
1,TCK-19268,Error al actualizar correo electrónico: bloque...,"Hola, buenas tardes. Escribo porque estoy trat...",784,Whatsapp,Otro,Premium,192,2024-01-02,25,Media
2,TCK-19269,Cobro desconocido en mi tarjeta digital,"Hola buenas, escribo porque me aparecio un des...",600,Whatsapp,Fraude,Free,314,2024-01-02,10,Baja
3,TCK-19270,"Cuenta bloqueada sin aviso, necesito ayuda urg...","Hola, necesito que me solucionen esto luego po...",659,Whatsapp,Cuenta,Free,228,2024-01-02,2,Media
4,TCK-19271,Me descontaron plata del cajero y no salió el ...,Hola buenos dias escribo porque ya estoy aburr...,853,Whatsapp,Otro,Free,72,2024-01-03,24,Media


Ahora, analiza los datos de entrada usando las herramientas solicitadas _(código: 1.0 pts en total)_:


1. Utiliza `.info()` y `.describe()` para obtener una visualización rápida del dataframe. _(0.1 pts)_
2. Utiliza `histogram` de plotly express para explorar la distribución de la variable objetivo, y `value_counts` más operaciones entre series para calcular la cantidad y porcentaje de sus categorías. _(0.2 pts)_
3. Utiliza `histogram` más [subplots](https://plotly.com/python/subplots/) para observar la distribución de características numéricas y categóricas. Hint: puedes acceder con `fig.data[0]` a la traza de un histograma generado por `histogram`. _(0.2 pts)_
4. Utiliza `corr` + `imshow` para analizar la correlación entre variables numéricas. _(0.1 pts)_
5. Genera bins con `qcut` para `N_Caracteres_Ticket`, `Tiempo_de_Resolución_Horas` y `Usuario-Antiguedad_Cuenta_Dias`, con los valores `Bajo`, `Medio`, `Alto`. Luego, utiliza `crosstab` (normalizando usando `normalize` por filas) + `imshow` para analizar las contingencias entre `Categoría_Problema` y `Nivel_Prioridad`, `Canal_Ticket` y `Nivel_Prioridad`, `Usuario-Tipo_de_Cuenta` y `Nivel_Prioridad`, y los bins de `N_Caracteres_Ticket` y `Nivel_Prioridad`, `Tiempo_de_Resolución_Horas` y `Nivel_Prioridad`, y `Usuario-Antiguedad_Cuenta_Dias` y `Nivel_Prioridad`. Grafica en subplots. _(0.2 pts)_
6. Por último, agrega la cantidad de reclamos por mes y `Nivel_Prioridad` (crea una nueva columna llamada `Fecha_Envío_Mes` usando el módulo de datetimes de pandas) para obtener un conteo de cantidad de reclamos según mes-nivel de prioridad, y utiliza `line` para generar un gráfico de línea en donde cada `Nivel_Prioridad` tenga su propia serie. _(0.2 pts)_

In [3]:
# Vistazo rápido
df_tickets.info()
df_tickets.describe()

<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column                          Non-Null Count  Dtype
---  ------                          --------------  -----
 0   Id_Ticket                       2000 non-null   str  
 1   Asunto_Ticket                   2000 non-null   str  
 2   Contenido_Ticket                2000 non-null   str  
 3   N_Caracteres_Ticket             2000 non-null   int64
 4   Canal_Ticket                    2000 non-null   str  
 5   Categoría_Problema              2000 non-null   str  
 6   Usuario-Tipo_de_Cuenta          2000 non-null   str  
 7   Usuario-Antiguedad_Cuenta_Dias  2000 non-null   int64
 8   Fecha_Envío                     2000 non-null   str  
 9   Tiempo_de_Resolución_Horas      2000 non-null   int64
 10  Nivel_Prioridad                 2000 non-null   str  
dtypes: int64(3), str(8)
memory usage: 1.6 MB


,N_Caracteres_Ticket,Usuario-Antiguedad_Cuenta_Dias,Tiempo_de_Resolución_Horas
count,2000.000000,2000.000000,2000.000000
mean,717.699000,121.743000,18.718000
std,209.355096,142.339366,18.841788
min,251.000000,4.000000,1.000000
25%,564.000000,42.000000,5.000000
50%,697.000000,78.000000,13.000000
75%,850.250000,143.000000,26.000000
max,1831.000000,1200.000000,183.000000


In [4]:
# Distribución de la variable objetivo
import plotly.express as px

fig = px.histogram(df_tickets, x="Nivel_Prioridad", title="Distribución de Nivel_Prioridad")
fig.show()

conteo = df_tickets["Nivel_Prioridad"].value_counts()
porcentaje = (df_tickets["Nivel_Prioridad"].value_counts(normalize=True) * 100).round(2)
pd.concat([conteo, porcentaje], axis=1, keys=["Cantidad", "Porcentaje (%)"])

,Cantidad,Porcentaje (%)
Nivel_Prioridad,,
Media,789,39.45
Baja,723,36.15
Alta,350,17.50
Critica,138,6.90


In [5]:
# Distribución de numéricas y categóricas en subplots
from plotly.subplots import make_subplots

NUM_COLS = ["N_Caracteres_Ticket", "Usuario-Antiguedad_Cuenta_Dias", "Tiempo_de_Resolución_Horas"]
CAT_COLS = ["Canal_Ticket", "Categoría_Problema", "Usuario-Tipo_de_Cuenta"]
cols = NUM_COLS + CAT_COLS

fig = make_subplots(rows=2, cols=3, subplot_titles=cols)
for i, col in enumerate(cols):
    hist = px.histogram(df_tickets, x=col)
    fig.add_trace(hist.data[0], row=i // 3 + 1, col=i % 3 + 1)
fig.update_layout(height=700, showlegend=False, title="Distribución de variables")
fig.show()

In [6]:
# Correlación entre numéricas
corr = df_tickets[NUM_COLS].corr()
fig = px.imshow(corr, text_auto=".2f", title="Correlación entre variables numéricas")
fig.show()

In [7]:
# Bins + contingencias con Nivel_Prioridad
BIN_LABELS = ["Bajo", "Medio", "Alto"]

df_tickets["N_Caracteres_Ticket_Bin"] = pd.qcut(df_tickets["N_Caracteres_Ticket"], q=3, labels=BIN_LABELS)
df_tickets["Tiempo_Resolucion_Bin"] = pd.qcut(df_tickets["Tiempo_de_Resolución_Horas"], q=3, labels=BIN_LABELS)
df_tickets["Antiguedad_Cuenta_Bin"] = pd.qcut(df_tickets["Usuario-Antiguedad_Cuenta_Dias"], q=3, labels=BIN_LABELS)

pares_contingencia = [
    ("Categoría_Problema", "Nivel_Prioridad"),
    ("Canal_Ticket", "Nivel_Prioridad"),
    ("Usuario-Tipo_de_Cuenta", "Nivel_Prioridad"),
    ("N_Caracteres_Ticket_Bin", "Nivel_Prioridad"),
    ("Tiempo_Resolucion_Bin", "Nivel_Prioridad"),
    ("Antiguedad_Cuenta_Bin", "Nivel_Prioridad"),
]

fig = make_subplots(rows=2, cols=3, subplot_titles=[f"{a} vs {b}" for a, b in pares_contingencia])
for i, (col_a, col_b) in enumerate(pares_contingencia):
    tabla = pd.crosstab(df_tickets[col_a], df_tickets[col_b], normalize="index")
    heatmap = px.imshow(tabla, text_auto=".2f")
    fig.add_trace(heatmap.data[0], row=i // 3 + 1, col=i % 3 + 1)
fig.update_layout(height=800, title="Contingencias vs Nivel_Prioridad")
fig.show()

In [8]:
# Reclamos por mes y nivel de prioridad
df_tickets["Fecha_Envío_Mes"] = pd.to_datetime(df_tickets["Fecha_Envío"]).dt.to_period("M").dt.to_timestamp()

conteo_mensual = df_tickets.groupby(["Fecha_Envío_Mes", "Nivel_Prioridad"]).size().reset_index(name="Cantidad")

fig = px.line(
    conteo_mensual,
    x="Fecha_Envío_Mes",
    y="Cantidad",
    color="Nivel_Prioridad",
    title="Cantidad de tickets por mes y nivel de prioridad",
)
fig.show()

Ahora, responde las siguientes preguntas _(preguntas: 1.6 pts en total)_:

1. ¿Cuántos datos son? ¿Hay nulos? _(0.1 pts)_
2. ¿Qué columnas tenemos disponibles? ¿Qué tipo de dato es cada columna? - Tanto en tipo de dato (str, int, float, etc.) como también en su contenido (texto libre, categorías, entero, número decimal, fechas, índices) y su relación. Responde en una tabla con: Columna, Tipo de dato, Tipo de contenido y Descripción. _(0.2 pts)_
3. ¿Cómo se distribuyen las etiquetas? ¿Hay desbalance? Incluye porcentajes en la respuesta. _(0.2 pts)_
4. ¿Cómo afecta la distribución de clases a las respuestas dadas en la sección 1? ¿Sería necesario editar o corregir algo de estas? Si es así, edita las respuestas de la sección 1 y menciona qué edición hiciste aquí. _(0.1 pts)_
5. ¿Cómo se distribuyen las variables numéricas? ¿Con cuáles distribuciones las podríamos relacionar? ¿Cuál es la interpretación de estas? ¿Existen outliers? Explica por cada variable numérica. _(0.2 pts)_
6. ¿Existe alguna correlación entre las variables numéricas? _(0.1 pts)_
7. ¿Existe alguna relación interesante entre las categorías generadas y las categorías de la variable objetivo? _(0.2 pts)_
8. ¿Existe alguna relación entre el tiempo y la variable objetivo? ¿Sería recomendable utilizar esta variable? _(0.1 pts)_
9. Pensemos ahora por un momento en el modelo en modo inferencia/producción. ¿Existe alguna variable que no vaya a estar disponible en ese momento? Si es así, ¿cómo se llama este fenómeno? ¿Si no hubiésemos meditado esto, qué hubiese sucedido? _(0.2 pts)_
10. Para generar los bins de las variables numéricas se utilizó `qcut` (cuantiles) en lugar de `cut` (ancho fijo). ¿Qué habría pasado con el heatmap de contingencia de `Tiempo_de_Resolución_Horas` si hubieras usado `cut`? ¿Por qué `qcut` es preferible cuando la variable está muy sesgada? _(0.2 pts)_

1. Son 2000 datos en total y no hay nulos, el .info() muestra las 11 columnas con "non-null", ninguna le falta un valor.

2. Tenemos Id_Ticket como identificador único de cada ticket. Asunto_Ticket y Contenido_Ticket son texto libre con el título y la descripción completa del problema. N_Caracteres_Ticket es un entero que mide el largo del contenido. Canal_Ticket, Categoría_Problema y Usuario-Tipo_de_Cuenta son categóricas, por dónde llegó el ticket, qué tipo de problema es, y qué plan tiene el usuario (Free, Premium, Business) respectivamente. Usuario-Antiguedad_Cuenta_Dias es un entero con los días desde que se creó la cuenta. Fecha_Envío es la fecha en que se envió el ticket. Tiempo_de_Resolución_Horas es un entero con las horas que tardó en resolverse. Y por último Nivel_Prioridad es nuestro target, una categórica (Baja, Media, Alta, Critica).

3. Hay desbalance claro, Media 39.45%, Baja 36.15%, Alta 17.50% y Critica solo 6.90%. La clase que más nos importa de negocio es justo la más chica.

4. Confirma que Accuracy sería una mala elección acá, ya que un modelo simple que solo prediga Media o Baja ya acertaría cerca del 75% de las veces sin aprender nada sobre Critica. Esto respalda la elección de macro recall/F1 hecha en la sección 1 en base a los supuestos que se confirmaron ahora, no se cambia nada por los supuestos tomados y confirmados acá.

5. Las tres numéricas están sesgadas a la derecha (cola larga hacia valores altos), ninguna se ve normal:
   - **N_Caracteres_Ticket**: la mayoría entre 500-900 caracteres, cola hasta 1831.
   - **Usuario-Antiguedad_Cuenta_Dias**: fuertemente concentrada cerca de 0 (usuarios recién llegados), cola hasta 1200 días.
   - **Tiempo_de_Resolución_Horas**: casi todo se resuelve bajo 50 horas, con pocos casos que se disparan hasta 183. A priori no parecen errores, podrían ser perfectamente casos reales (clientes muy antiguos, tickets excepcionalmente demorados), así que no los trataría como outliers a eliminar, sino como cola natural del negocio.

6. Prácticamente no hay correlación entre las numéricas, la más alta es 0.19 entre N_Caracteres_Ticket y Tiempo_de_Resolución_Horas, el resto está entre 0.02 y 0.07. Concluimos que son variables bastante independientes entre sí.

7. Varias relaciones tienen sentido de negocio:
   - Fraude es la categoría con más Critica (24%) y más Alta (37%), lógico dado que involucra dinero.
   - Pregunta general tiene 0% en Alta y Critica, coherente con ser consultas no urgentes.
   - Business tiene 0% Baja y 63% combinado entre Alta y Critica, los usuarios de mayor valor generan tickets más urgentes.
   - El patrón más claro es con N_Caracteres_Ticket, donde los tickets cortos son 68% Baja, mientras los largos concentran más Alta y Critica (52% combinado). Lógico donde la gente escribe más cuando el problema es grave.
   - Cuentas nuevas (antigüedad baja) tienen casi el doble de Alta+Critica que cuentas antiguas.

8. No se ve una tendencia tan clara ni estacionalidad fuerte, donde las series de Baja, Media y Alta fluctúan con ruido mes a mes durante los dos años, con Alta mostrando un leve aumento hacia el final. La caída brusca en enero 2026 es porque ese mes está incompleto en los datos, no una baja real. No lo usaría como un feature fuerte, no muestra un patrón sistemático aprovechable.

9. Sí, `Tiempo_de_Resolución_Horas`. Este dato solo se conoce después de que el ticket ya fue resuelto, pero el modelo debe predecir la prioridad apenas llega el ticket, cuando ese tiempo todavía no existe. Esto es lo que se llama data leakage. Si no lo hubiéramos detectado, el modelo habría aprendido a depender de esta variable (dado que tickets urgentes probablemente se resuelven distinto a los no urgentes) y mostraría métricas artificialmente buenas en el entrenamiento, pero fallaría por completo en producción porque esa columna simplemente no existe al momento de clasificar un ticket nuevo.

10. Con `cut` (ancho fijo) sobre una variable tan sesgada como Tiempo_de_Resolución_Horas, casi todos los tickets caerían en el primer bin, como el 75% de los datos está bajo 26 horas y el máximo es 183, un ancho fijo dejaría bins como "Medio" y "Alto" casi vacíos. El heatmap resultante tendría celdas con poquísimas observaciones, haciendo las proporciones poco confiables y el gráfico prácticamente inútil para comparar. `qcut` evita esto porque arma los bins por cuantiles, asegurando una cantidad similar de tickets en cada uno, lo que da comparaciones estables incluso con una variable muy sesgada.

---

## Subsección 2 - Embeddings

_(Total: 1.4 pts — código 0.6 + preguntas 0.8)_

En esta sección trabajarás con embeddings sobre los Asuntos + Descripciones de tickets precalculados con el modelo `gemini-embeddings-001`. Ahora, nuevamente utilizando las herramientas de lectura de parquet _(código: 0.6 pts en total)_:

1. Lee el archivo que contiene los embeddings (`data/embeddings.parquet`). _(0.1 pts)_
2. Reduce la dimensionalidad de los embeddings con umap-learn a dos dimensiones y agrega a `df_tickets` las nuevas dimensiones. Asegura tu proyección utilizando semillas, así luego, al momento de la revisión, se podrán replicar los resultados. _(0.3 pts)_
3. Grafica un scatter plot con las proyecciones y pinta por `Nivel_Prioridad` (utiliza el código dado).
4. Grafica un nuevo scatter plot con las proyecciones y pinta por `Categoría_Problema`. _(0.2 pts)_

In [ ]:
# Lectura de embeddings.parquet
df_embeddings = pd.read_parquet("data/embeddings.parquet")
df_embeddings.head()

,embedding_dim_1,embedding_dim_2,embedding_dim_3,embedding_dim_4,embedding_dim_5,embedding_dim_6,embedding_dim_7,embedding_dim_8,embedding_dim_9,embedding_dim_10,...,embedding_dim_1015,embedding_dim_1016,embedding_dim_1017,embedding_dim_1018,embedding_dim_1019,embedding_dim_1020,embedding_dim_1021,embedding_dim_1022,embedding_dim_1023,embedding_dim_1024
Id_Ticket,,,,,,,,,,,,,,,,,,,,,
TCK-19267,-0.011585,-0.008728,0.012878,-0.050069,0.006369,0.008990,-0.008538,0.010781,-0.021699,0.007514,...,-0.011539,-0.016249,0.004422,-0.022274,0.004511,0.001428,-0.007812,-0.020485,-0.001888,-0.000408
TCK-19268,0.014494,0.004070,0.011298,-0.056227,0.009839,0.014777,0.005187,0.001136,-0.000801,-0.006830,...,-0.019967,-0.008424,-0.002084,-0.002761,-0.004148,-0.007131,-0.012717,-0.019045,-0.015172,0.004308
TCK-19269,-0.010936,0.020796,0.011680,-0.080212,0.017745,0.020321,0.004509,0.000413,-0.009146,0.008450,...,-0.021763,-0.007094,-0.011453,-0.001084,-0.005820,-0.015085,0.003200,-0.019789,-0.018502,-0.001085
TCK-19270,0.003248,0.005420,0.003422,-0.065296,0.026466,0.018159,-0.011115,0.004692,0.001486,-0.006940,...,-0.019830,-0.007490,-0.013525,0.000086,-0.011983,0.007148,-0.002536,-0.016200,-0.003927,-0.004028
TCK-19271,-0.010437,0.013634,0.009859,-0.066499,0.006272,0.006100,-0.010604,0.009686,-0.032323,-0.003181,...,-0.006401,-0.001971,0.002344,-0.010598,-0.011034,-0.012039,-0.000857,-0.009420,-0.006597,0.012463


In [ ]:
# Reducción a 2D con UMAP (semilla fija para reproducibilidad: mi rut)
import umap

reducer = umap.UMAP(n_components=2, random_state=20968061)
embedding_2d = reducer.fit_transform(df_embeddings.values)

df_embeddings["x"] = embedding_2d[:, 0]
df_embeddings["y"] = embedding_2d[:, 1]

df_tickets = df_tickets.merge(df_embeddings[["x", "y"]], left_on="Id_Ticket", right_index=True, how="left")

/Users/tomasvallejo/Documents/Universidad/Otoño 2026/MDS7202/.venv/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/Users/tomasvallejo/Documents/Universidad/Otoño 2026/MDS7202/.venv/lib/python3.14/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(


In [11]:
import textwrap


def wrap_text(text, width=40):
    return "<br>".join(textwrap.wrap(text, width=width))


df_tickets["Contenido_Ticket_Wrapped"] = df_tickets["Contenido_Ticket"].apply(lambda x: wrap_text(x, 80))

fig_proj = px.scatter(
    df_tickets,
    x="x",
    y="y",
    color="Nivel_Prioridad",
    title="Proyecciones de los embeddings respecto a clase",
    hover_data=[
        "Canal_Ticket",
        "N_Caracteres_Ticket",
        "Categoría_Problema",
        "Usuario-Tipo_de_Cuenta",
        "Usuario-Antiguedad_Cuenta_Dias",
        "Contenido_Ticket_Wrapped",
    ],
)
fig_proj.show()

In [12]:
fig_cat = px.scatter(
    df_tickets,
    x="x",
    y="y",
    color="Categoría_Problema",
    title="Proyecciones de los embeddings respecto a categoría de problema",
    hover_data=[
        "Canal_Ticket",
        "N_Caracteres_Ticket",
        "Nivel_Prioridad",
        "Usuario-Tipo_de_Cuenta",
        "Usuario-Antiguedad_Cuenta_Dias",
        "Contenido_Ticket_Wrapped",
    ],
)
fig_cat.show()

Ahora, responde las siguientes preguntas _(preguntas: 0.8 pts en total)_:

1. ¿Qué es un embedding? ¿Qué representan? A grandes rasgos, ¿cómo se calculan? Hint: consulta sobre `gemini-embedding-001`. _(0.2 pts)_
2. ¿Por qué los embeddings tienen el potencial de ser mejores para un modelo predictivo que analizar las palabras individuales? _(0.1 pts)_
3. ¿Cuál podría ser una desventaja de utilizar embeddings en vez de utilizar directamente palabras? _(0.1 pts)_
4. ¿Existen otros tipos de embeddings? ¿Qué podrían representar? _(0.1 pts)_
5. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Nivel_Prioridad`? _(0.1 pts)_
6. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Categoría_Problema`? _(0.1 pts)_
7. Las proyecciones de UMAP (`x`, `y`) quedaron agregadas a `df_tickets`. ¿Sería buena idea usar estas 2 dimensiones como features de entrada del modelo predictivo, en lugar de los embeddings originales? Justifica pensando en la información que se pierde al reducir a 2D. _(0.1 pts)_

1. Un embedding es una representación numérica de un texto en forma de vector, donde textos con significado parecido quedan ubicados cerca entre sí en ese espacio vectorial. `gemini-embedding-001` los calcula pasando el texto por un modelo transformer entrenado específicamente para que la distancia entre vectores refleje similitud semántica, no similitud de palabras exactas.

2. Porque capturan significado y no solo la palabra literal. Un ticket que dice "no puedo entrar a mi cuenta" y otro que dice "perdí acceso a mi perfil" usan palabras completamente distintas pero significan lo mismo, y en el espacio de embeddings quedarían cerca. Analizar palabra por palabra (como Bag-of-Words) no captura esa relación, cada palabra distinta es un feature separado sin conexión entre sí.

3. Se pierde interpretabilidad, ya que con palabras uno puede mirar directamente qué término influyó en la predicción ("urgente", "fraude", etc.), pero con un embedding cada dimensión es un número abstracto que no tiene un significado claro, así que es mucho más difícil explicarle a alguien del negocio por qué el modelo decidió algo.

4. Sí, existen embeddings de imágenes (representan contenido visual), de usuarios en sistemas de recomendación (representan gustos o comportamiento), de grafos (representan relaciones entre nodos), y embeddings multimodales como los de `gemini-embedding-2`, que mapean texto, imagen, audio y video a un mismo espacio vectorial compartido.

5. Se ven dos grandes agrupaciones separadas en el eje x, pero dentro de cada agrupación los cuatro niveles de prioridad están bastante mezclados entre sí, sin una separación tan clara, aunque se puede llegar a notar una separabilidad entre las dos clases extremo donde prioridad "baja" la mayoría quedan a la izquierda del eje x, mientras que las de prioridad "critica" quedan a la derecha. Esta baja separación sugiere que el contenido semántico del ticket por sí solo no separa bien la prioridad, probablemente porque la urgencia depende más del contexto del usuario que de las palabras usadas.

6. Acá sí se puede observar un patrón más claro, donde el grupo de la derecha está mayormente dominado por tickets de `Fraude` y `Cobros`, mientras que el grupo de la izquierda concentra `Cuenta`, `Técnica`, `Otro` y `Pregunta general`. Esto tiene sentido, el contenido semántico de un ticket sí distingue bien de qué tipo de problema se trata, aunque no distinga qué tan urgente es.

7. No, no sería buena idea. Reducir 1024 dimensiones a solo 2 para UMAP implica perder muchísima información en el proceso, UMAP prioriza preservar la estructura local de vecinos cercanos para poder graficar, no preservar toda la información útil para clasificar. Los patrones que vimos muestran que la reducción sirve para visualizar agrupaciones generales, pero perdería muchos matices finos que el modelo sí podría aprovechar si usa las 1024 dimensiones originales completas.